src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs

```diff
@@ -2,57 +2,82 @@
+        {
+            if (data == null) throw new ArgumentNullException(nameof(data));
+            if (offset < 0) throw new ArgumentOutOfRangeException(nameof(offset), SR.ArgumentOutOfRange_NeedNonNegNum);
+            if (count < 0) throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_NeedNonNegNum);
+            if (offset + count > data.Length) throw new ArgumentException(SR.Argument_InvalidOffLen);
+
+            GetBytesInternal(data, offset, count);
+        }
 
-            int indexOfFirst0Byte = data.Length;
-            for (int i = 0; i < data.Length; i++)
+        public override void GetNonZeroBytes(byte[] data)
+        {
+            if (data == null) throw new ArgumentNullException(nameof(data));
+
+            int offset = 0;
+            while (offset < data.Length)
             {
-                if (data[i] == 0)
+                // Fill the remaining portion of the array with random bytes.
+                GetBytesInternal(data, offset, data.Length - offset);
+
+                // Find the first zero in the remaining portion.
+                int indexOfFirst0Byte = data.Length;
+                for (int i = offset; i < data.Length; i++)
                 {
-                    indexOfFirst0Byte = i;
-                    break;
+                    if (data[i] == 0)
+                    {
+                        indexOfFirst0Byte = i;
+                        break;
+                    }
                 }
-            }
 
-            for (int i = indexOfFirst0Byte; i < data.Length; i++)
-            {
-                if (data[i] != 0)
+                // If there were any zeros, shift down all non-zeros,
+                // leaving all of the zeros at the end.
+                for (int i = indexOfFirst0Byte; i < data.Length; i++)
                 {
-                    data[indexOfFirst0Byte++] = data[i];
+                    if (data[i] != 0)
+                    {
+                        data[indexOfFirst0Byte++] = data[i];
+                    }
                 }
+
+                // Go again on the space filled with zeros, if any.
+                offset = indexOfFirst0Byte;
             }
+        }
 
-            while (indexOfFirst0Byte < data.Length)
-            {
-                // This should be more than enough to fill the rest in one iteration
-                byte[] tmp = new byte[2 * (data.Length - indexOfFirst0Byte)];
-                GetBytes(tmp);
+        private void GetBytesInternal(byte[] data, int offset, int count)
+        {
+            Debug.Assert(data != null);
+            Debug.Assert(offset >= 0);
+            Debug.Assert(count >= 0);
+            Debug.Assert(offset + count <= data.Length);
 
-                for (int i = 0; i < tmp.Length; i++)
+            if (count > 0)
+            {
+                unsafe
                 {
-                    if (tmp[i] != 0)
+                    fixed (byte* pb = &data[offset])
```

Comment:
I have a test for it :) ... I found this on msdn (see &arr[5] example) https://msdn.microsoft.com/en-us/library/f58wzh21.aspx although for bytes just doing a + would be pretty standard I guess